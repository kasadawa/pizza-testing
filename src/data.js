export var pizzas = [{
    name:'Margarita',
    image: 'margarita.jpg',
    quantity: 1, 
    descr:'tomato, and other ingredience',
    style:'normal',
    options: [],
},{
    name:'Devil pizza',
    image: 'deliv.jpg',
    quantity: 1,
    descr:'ham,tomato, and other ingredience',
    style:'normal',
    options: []
},{
    name:'Calcone',
    descr:'mushrooms,ham,tomato, and other ingredience',
    quantity: 1,
    image: 'calcone.jpg',
    style:'normal',
    options: []
}];

export var prices = new Map();
prices.set('Margarita',12) 
prices.set('Devil pizza',14.4) 
prices.set('Calcone',15)
prices.set('italian', 2) 
prices.set('exstra cheese', 1.3) 
prices.set('exstra tomato', 3)


export function cloneArr(arr){
    return arr.map((val)=>{
        if(typeof val === 'object'){
            return cloneObj(val);
        }
    })
}

export function cloneObj(obj){
    var copy = {} ; 
    for(let key in obj){
        if(obj.hasOwnProperty(key)){
            if(typeof obj[key] === 'object' && !Array.isArray(obj[key])){
                copy[key] = cloneObj(obj[key]);
            }else if( Object.prototype.toString.call(obj[key]) === '[object Array]'){
                copy[key] =[];
            }else{
                copy[key] = obj[key];
            }
        }
    }
    return copy; 
}