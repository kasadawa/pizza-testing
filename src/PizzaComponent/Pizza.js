import  React from 'react';
import { pizzas, cloneObj } from '../data';


import margarita  from '../img/margarita.jpg';
import devil  from '../img/devil.jpg';
import calcone  from '../img/calcone.jpg';


import './Pizza.css';


// var orderPizzaObj = {
//     extraTomato: Bool, 
//     extraCheese: Bool,
//     pizzaName:String, 
//     totalPrice: Number,
//     type:String,
//     quantity: Number, 
// }



[pizzas[0].image, pizzas[1].image,pizzas[2].image] = [margarita,devil,calcone];

export default class Pizza extends React.Component {

    constructor(props){
        super(props);
        this.state = cloneObj(pizzas[this.props.index]);


        //bindings 
        this.changeStyle = this.changeStyle.bind(this);
    }


    changeQuantity(type){ 
        if(!isNaN(this.state.quantity) && !(this.state.quantity === 1 && type ==='dec') ){
           
            this.setState((state)=>{
                state.quantity =  type === 'inc' ? state.quantity+ 1 : state.quantity- 1; 
                return state;
            })
        }
    }

    changeStyle(e){
        this.setState({style:e.target.value});
    }


    changeIngredient(e){
        if(e.target.checked){
            this.state.options.push(e.target.value);
        }else{
            let i = this.state.options.indexOf(e.target.value);
            this.state.options.splice(i,1);
        }
        this.setState(this.state);
    }



    addToBasket(){
        var orderNum = window.sessionStorage.getItem('ordernum'); 
        // +null = 0 :O 
        if(isNaN(+orderNum) || orderNum == null ){ // if someone changed the sessionStorage or its null
            window.sessionStorage.setItem('ordernum',1);
            orderNum = 1 ;
        }else{
            orderNum++;
        }
        window.sessionStorage.setItem('order' + orderNum ,JSON.stringify(this.state));
        window.sessionStorage.setItem('ordernum',orderNum);
        document.getElementById('ordernum').innerText = orderNum;
        this.setState(cloneObj(pizzas[this.props.index]));
        
        // clear selected input fields
        document.getElementById('tomato'+this.props.index).checked = false; 
        document.getElementById('cheese'+this.props.index).checked = false; 
        
    }

  render() {
    return (
      <div>    
            <div className="media">
              <img src={this.state.image} alt='' className="mr-3 resize" />
              <div className="media-body">
                <h5 className="mt-0">{this.props.name}</h5>
                Ingredients: {this.state.descr}

                <div className="col-md-4">
                <p>Quantity <span>{this.state.quantity}</span><i onClick={this.changeQuantity.bind(this,'inc')} className="fas fa-plus"></i><i onClick={this.changeQuantity.bind(this,'dec')} className="fas fa-minus"></i></p>
                </div>
                
               <div className='row'>
                <div className="col-md-4">
                    <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <label className="input-group-text" htmlFor={"inputGroupSelect0" + this.props.index}>Pizza Style</label>
                    </div>
                    <select value={this.state.style} onChange={(e)=>this.changeStyle(e)} className="custom-select" id={"inputGroupSelect0" + this.props.index}>
                        <option value="clasic" selected>clasic</option>
                        <option value="italian">italian</option>
                    </select>
                    </div>
                </div>
                <div className="col-md-4"></div>
                 
                <div className="col-md-4">
                    <button onClick={this.addToBasket.bind(this)} className="btn btn-success" >Add to basket</button>
                </div>
                </div>
                <div className="col-md-4">

                <div className="input-group mb-3">
           

                    <div className="form-check">
                        <input onChange={(e)=>this.changeIngredient(e)} className="form-check-input" type="checkbox" value="cheese" id={"cheese"+ this.props.index }/>
                        <label className="form-check-label"  htmlFor={ "cheese" + this.props.index}>
                            Extra Cheese
                        </label>
                        <br></br>
                        <input onChange={(e)=>this.changeIngredient(e)} className="form-check-input" type="checkbox" value="tomato" id={"tomato"+ this.props.index} />
                        <label className="form-check-label" htmlFor={"tomato" + this.props.index}>
                            Extra Tomato
                        </label>
                    </div>

                    </div>
                  
                </div> 

              </div> 

            </div>
      </div>
    );
  }
}