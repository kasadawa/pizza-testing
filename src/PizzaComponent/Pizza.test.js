import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, {shallow,render, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {pizzas} from '../data';

Enzyme.configure({adapter: new Adapter()});

import Pizza from './Pizza';


test('Pizza component snapshot', () => {
  const component = renderer.create(
    <Pizza></Pizza>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});

describe('Testing Pizza component',()=>{

    
    it('increment pizza', () => {
        
        for(let i = 0 ; i < pizzas.length; i++ ){
        const wrapper = shallow(<Pizza index={i} />);
        wrapper.find('i.fas.fa-plus').simulate('click');
        expect(wrapper.state().quantity).toBe(2);         
    }
  });



  it('decrement pizza', () => {
    
    for(let i = 0 ; i < pizzas.length; i++ ){
        const wrapper = shallow(<Pizza index={i} />);
        
        // do not decrement if the quantity is 1 
        expect(wrapper.state().quantity).toBe(1); 
        wrapper.find('i.fas.fa-minus').simulate('click');
        expect(wrapper.state().quantity).toBe(1);  
        
        // decrement if quantity is 2 
        wrapper.find('i.fas.fa-plus').simulate('click'); 
        expect(wrapper.state().quantity).toBe(2);  
        wrapper.find('i.fas.fa-minus').simulate('click');
        expect(wrapper.state().quantity).toBe(1); 
    }
  });



});

 

