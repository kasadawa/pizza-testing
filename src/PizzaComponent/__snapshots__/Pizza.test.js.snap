// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pizza component snapshot 1`] = `
<div>
  <div
    className="media"
  >
    <img
      alt=""
      className="mr-3 resize"
    />
    <div
      className="media-body"
    >
      <h5
        className="mt-0"
      />
      Ingredients: 
      <div
        className="col-md-4"
      >
        <p>
          Quantity 
          <span />
          <i
            className="fas fa-plus"
            onClick={[Function]}
          />
          <i
            className="fas fa-minus"
            onClick={[Function]}
          />
        </p>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <div
            className="input-group mb-3"
          >
            <div
              className="input-group-prepend"
            >
              <label
                className="input-group-text"
                htmlFor="inputGroupSelect0undefined"
              >
                Pizza Style
              </label>
            </div>
            <select
              className="custom-select"
              id="inputGroupSelect0undefined"
              onChange={[Function]}
            >
              <option
                selected={true}
                value="clasic"
              >
                clasic
              </option>
              <option
                value="italian"
              >
                italian
              </option>
            </select>
          </div>
        </div>
        <div
          className="col-md-4"
        />
        <div
          className="col-md-4"
        >
          <button
            className="btn btn-success"
            onClick={[Function]}
          >
            Add to basket
          </button>
        </div>
      </div>
      <div
        className="col-md-4"
      >
        <div
          className="input-group mb-3"
        >
          <div
            className="form-check"
          >
            <input
              className="form-check-input"
              id="cheeseundefined"
              onChange={[Function]}
              type="checkbox"
              value="cheese"
            />
            <label
              className="form-check-label"
              htmlFor="cheeseundefined"
            >
              Extra Cheese
            </label>
            <br />
            <input
              className="form-check-input"
              id="tomatoundefined"
              onChange={[Function]}
              type="checkbox"
              value="tomato"
            />
            <label
              className="form-check-label"
              htmlFor="tomatoundefined"
            >
              Extra Tomato
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
